Create a separate reference page for the  Processing engine (#5958)
influxdata/docs-v2: content/shared/influxdb3-processing-engine.md

# Processing engine

> The Processing engine is an embedded Python virtual machine that runs inside the InfluxDB database. It executes Python code in response to database events without requiring external application servers or middleware.

Reviewer: ```suggestion
The Processing engine is an embedded Python virtual machine that runs inside an {{% product-name %}} database server. It executes Python code in response to triggers and database events without requiring external application servers or middleware.
```

## How it works

### Architecture

> The Processing engine runs Python code directly within the InfluxDB process. This design provides high performance and direct access to database resources.

Reviewer: ```suggestion
The Processing engine runs Python code directly within a {{% product-name %}} server process. This design provides high performance and direct access to database resources.
```

> - **Embedded execution**: Code runs in the same process space as the database

Reviewer: ```suggestion
- **Embedded execution**: Code runs in the same process space as the database server
```

> - **Direct data access**: Zero-copy access to data where possible

Reviewer: ```suggestion
- **Direct data access**: Zero-copy access to data
```

- **Event-driven**: Responds to database writes, scheduled events, and HTTP requests
- **Isolated contexts**: Maintains separation between different plugin executions

> 

Reviewer: You could add a bullet that mentions access to the Distinct values and Last values caches, which are very important features.

### Event processing flow

When events occur in the database, the Processing engine handles them through a consistent sequence:

> 1. A **trigger** activates when a specific event occurs

Reviewer: This isn't quite how the trigger relationship works.

2. The engine loads the associated **plugin** (Python code)
3. The plugin receives relevant **context and data**

> 4. The code processes the data and can write results back

Reviewer: ```suggestion
4. The code can process received data, query data, call third party tools, and write data to the database (for example, downsampled data) and caches.
```

5. Execution completes and the engine returns to waiting state

## Key components

### Trigger system

Triggers connect database events to Python code execution based on specific conditions:

> - **Data write triggers**: Execute when data is written to specific tables

Reviewer: ```suggestion
- **Data write triggers**: Execute on WAL flush events, when data is written to the object store, for a specific table or all tables in a database 
```

Make sure the names are consistent with what we have elsewhere.

# Processing engine

The Processing engine is an embedded Python virtual machine that runs inside an {{% product-name %}} database server. It executes Python code in response to triggers and database events without requiring external application servers or middleware.

## How it works

### Architecture

The Processing engine runs Python code directly within a {{% product-name %}} server process. This design provides high performance and direct access to database resources.

- **Embedded execution**: Code runs in the same process space as the database server
- **Direct data access**: Zero-copy access to data
- **Event-driven**: Responds to database writes, scheduled events, and HTTP requests

> - **Isolated contexts**: Maintains separation between different plugin executions

Reviewer 2: I think we should be very careful with the wording here since we don't have true isolation. Consider:

```
$ cat ./plugins/print-sys.py
import importlib.util

def _tst_sys(influxdb3_local):
    try:
        import sys
        influxdb3_local.info("_tst_sys(): sys.path = %s" % sys.path)
    except Exception as e:
        influxdb3_local.error("_tst_sys(): %s" % e)
        return

def process_scheduled_call(influxdb3_local, call_time, args=None):
    # show paths
    _tst_sys(influxdb3_local)

$ cat ./plugins/update-sys.py
import importlib.util

def _tst_sys(influxdb3_local):
    try:
        import sys
        influxdb3_local.info("_tst_sys(): sys.path = %s" % sys.path)
    except Exception as e:
        influxdb3_local.error("_tst_sys(): %s" % e)
        return

def process_scheduled_call(influxdb3_local, call_time, args=None):
    # update sys.path
    import sys
    sys.path.append("/foo")

    # show paths
    _tst_sys(influxdb3_local)
```

`print-sys.py` only prints `sys.path` while `update-sys.py` updates it. Notice:
```
# print-sys.py - sys.path doesn't have '/foo'
$ ~/.influxdb/influxdb3 test schedule_plugin --database test print-sys.py
{
  "trigger_time": "2025-04-23T17:53:25Z",
  "log_lines": [
    "INFO: _tst_sys(): sys.path = ['/home/jamie/.influxdb/python/lib/python313.zip', '/home/jamie/.influxdb/python/lib/python3.13', '/home/jamie/.influxdb/python/lib/python3.13/lib-dynload', '/tmp/data00/plugins/.venv/lib/python3.13/site-packages']"
  ],
  "database_writes": {},
  "errors": []
}

# update-sys.py to add '/foo' to sys.path
$ ~/.influxdb/influxdb3 test schedule_plugin --database test update-sys.py
{
  "trigger_time": "2025-04-23T17:54:21Z",
  "log_lines": [
    "INFO: _tst_sys(): sys.path = ['/home/jamie/.influxdb/python/lib/python313.zip', '/home/jamie/.influxdb/python/lib/python3.13', '/home/jamie/.influxdb/python/lib/python3.13/lib-dynload', '/tmp/data00/plugins/.venv/lib/python3.13/site-packages', '/foo']"
  ],
  "database_writes": {},
  "errors": []
}

# print-sys.py - sys.path has '/foo'! NOT ISOLATED
$ ~/.influxdb/influxdb3 test schedule_plugin --database test print-sys.py
{
  "trigger_time": "2025-04-23T17:54:58Z",
  "log_lines": [
    "INFO: _tst_sys(): sys.path = ['/home/jamie/.influxdb/python/lib/python313.zip', '/home/jamie/.influxdb/python/lib/python3.13', '/home/jamie/.influxdb/python/lib/python3.13/lib-dynload', '/tmp/data00/plugins/.venv/lib/python3.13/site-packages', '/foo']"
  ],
  "database_writes": {},
  "errors": []
}
```

There are probably many ways a plugin could interfere with another plugin with the current implementation.

cc @pauldix and @hiltontj for awareness

Author: Thank you for the information. I made a revision to more accureatly reflect plugin executions. Would you checking over it again?

- **Cache integration**: Access to system caches including Last values  and Distinct values 

### Event processing flow

When events occur in the database, the Processing engine handles them through a consistent sequence:

> 1. A **trigger** activates the plugin based on one of three event types:

Reviewer: ```suggestion
1. A **trigger** specific to one of the event types activates its plugin. The event types include:
```

   - Data writes to specific tables or all tables
   - Scheduled events (time-based or cron expressions)
   - HTTP requests to configured endpoints
2. The engine loads the associated **plugin** specified in the trigger configuration
3. The plugin receives context data specific to the trigger type:
   - Write triggers: the written data and table information
   - Schedule triggers: the scheduled call time
   - HTTP triggers: the request object with methods, headers, and body

> 4. The plugin processes the received data, can query the database, call external tools, and write results back

Author: Oh,  my apologies! That was a mistake. Thank you for catching that, I will revert it back.

The Processing engine is an embedded Python virtual machine that runs inside an {{% product-name %}} database server. It executes Python code in response to triggers and database events without requiring external application servers or middleware.

## How it works

### Architecture

The Processing engine runs Python code directly within a {{% product-name %}} server process. This design provides high performance and direct access to database resources.

- **Embedded execution**: Code runs in the same process space as the database server
- **Direct data access**: Zero-copy access to data
- **Event-driven**: Responds to database writes, scheduled events, and HTTP requests
- **Cache integration**: Access to system caches including Last values  and Distinct 
values 

> [!Note]

> > The Processing engine runs all plugins in the same Python process. Changes made by one plugin can affect other plugins, 
> > as the system does not fully isolate changes from each other. 

Reviewer: ```suggestion
> The Processing engine runs all plugins in the same Python process. Changes made by one plugin can affect other plugins.
```

I just noticed that the end of this sentence isn't correct, and probably isn't necessary unless we have more specifics.

### Event processing flow

> When events occur in the database, the Processing engine handles them through a consistent sequence:

Reviewer: ```suggestion
When specific events occur in the database, the Processing engine handles them through a consistent sequence:
```

> 1. A **trigger** specific to one of the event types activates its plugin. The event types include:

Reviewer: ```suggestion
1. A **trigger** specific to the event type activates its plugin. The event types include:
```

- **Resource limits**: Controls memory usage and execution time

## Performance characteristics

The Processing engine is designed for high-performance operation with minimal overhead:

- **Low latency**: Activates triggers in sub-millisecond time
- **Efficient access**: Accesses database directly without network overhead
- **Controlled resources**: Limits memory and CPU usage through configuration
- **Execution policies**: Offers synchronous or asynchronous processing options

## Reliability features

The Processing engine includes multiple features to ensure consistent and dependable execution:

- **Error handling**: Configures behaviors for failure scenarios (log, retry, or disable)
- **Execution tracking**: Tracks plugin performance and resource usage
- **State persistence**: Persists cache state across server restarts

## Extension capabilities

Extend and customize the Processing engine through several built-in mechanisms:

- **Package management**: Installs custom Python dependencies
- **Plugin distribution**: Distributes plugins via Git repositories
- **Shared API**: Provides consistent interface for database operations

> For a step-by-step guide to setting up and using the Processing engine, see the [Getting started with plugins](/influxdb3/core/plugins/) documentation.

Reviewer: ```suggestion
For a step-by-step guide to setting up and using the Processing engine, see the [Getting started with plugins](/influxdb3/version/plugins/) documentation.
```
